[Raptor] Lex walking duration

pbench opened the pull request:
Update the raptor algorithm to remember the walking duration corresponding to the earliest arrival time found.

More precisely, the raptor algorithm computes for each pair (_stop_point_, _nb_of_transfer_) the earliest arrival time at _stop_point_  using _nb_of_transfer_  change of vehicle ([example](https://github.com/CanalTP/navitia/blob/dev/documentation/rfc/doc-routing-algorithm.md#the-first-pass-on-example-1)).
Now, it also stores the walking duration of a journey that achieves this earliest arrival time.

The first commits are logs tracing the behavior of raptor.
The actual modifications of the algorithms are  :

- https://github.com/CanalTP/navitia/commit/297e405d0e8635de62990917088e73a4928e2089 adds storage to raptor Labels to remember the walking duration and updates this walking duration when the arrival time is updated for a pair (_stop_point_, _nb_of_transfer_)

- https://github.com/CanalTP/navitia/commit/6fe7efb250fd81e7dcc171fdc422a2c5c9bc5757 
The raptor algorithm works in "rounds". At round _k_ the earliest arrival time at every stop point using at most _k-1_ transfers is already computed, and the round will compute the earliest arrival time at every stop point using _k_ transfers.
In order to do so, raptor scans each _journey_pattern_ :

  - starting from the _first_stop_point_ of the _journey_pattern_, we embark on the first _vehicule_journey_ that we can catch from the earliest arrival time at _first_stop_point_ using _k-1_ transfers.
  - we "ride" the _vehicle_journey_, which gives new arrival times at each _stop_point_ along the ride. When this yields a better arrival time for a _stop_point_ we update its best known arrival time using _k_ transfers
  - it may happen that with our current ride, we arrive at a _stop_point_ later than the arrival time using _k-1_ transfers. In this case, we check if the arrival time using _k-1_ transfers allows to catch an earlier _vehicle_journey_ for this _journey_pattern_. When this happens, we "leave" the current _vehicle_journey_, embark on the earlier one, and remember the current stop_point as the "boarding" stop point.
Before the commit, the boarding stop_point was updated even if the "earlier vehicle journey" was the same as the current one. 
This commit updates the boarding stop_point if : 
    - it allows to catch a strictly earlier _vehicle_journey_
    - or it allows to catch the same _vehicle_journey_ but with a strictly smaller walking duration

- https://github.com/CanalTP/navitia/commit/64de81c2021043ab47396ea972b0bad3e4893aca 
the _fallback_duration_ of a StartingPointSecondPhase was initialized with the walking duration at the arrival stop point before https://github.com/CanalTP/navitia/commit/297e405d0e8635de62990917088e73a4928e2089 . After, _fallback_duration_ is initialized with the total walking duration.
In solution reader, the _fallback_duration_ of a StartingPointSecondPhase was used as the walking duration at the arrival stop point to build bound journeys. Now, bound journeys use the walking duration at arrival stop point computed before raptor (which is stored in a map_stop_point_duration) rather than _fallback_duration

**do_not_merge** for now as it may modify the output of navitia on several functional tests. More modifications may be needed before merge, and tests may need updates.
- tests on auvergne have been analyzed. When the output changed, it seems to yield better journeys than before. Other tests need to be analyzed.
- routing/raptor_test failed : to be analyzed

**JIRA** https://jira.kisio.org/browse/NAVITIAII-2814

Commit:
logs tracing raptor

Commit:
benchmark : add option to dump demands

Commit:
logs tracing label updates in raptor

Commit:
remember walking duration in raptor loop

Commit:
log walking duration in label updates

Commit:
update condition for updating boarding stop point

before : update if arrival time is better or equal
after : update if
- arrival time is strictly better
- arrival time is the same, and walking time is strictly better

Commit:
correct street network duration in bound journey

Commit:
correct sn_dur in fake journey  before second pass

Stop adding connection duration to sn_dur :
starting_point_second_phase.fallback_dur now includes the connection
durations. Hence, adding agin connection duration in
convert_to_bound count these durations twice.

Commit:
Update boarding stop point if arrival time and walking time are the same

fix raptor_test begin_different_zone2

Commit:
cleaning

removed comments, satisfy linters

Commit:
fix build and more cleaning

kinnou02 commented on source/routing/raptor.h:
```diff
@@ -201,7 +201,9 @@ struct RAPTOR {
                             const uint16_t l_zone,
-                            DateTime workingDate);
+                            DateTime workingDate,
+                            DateTime working_walking_duration,
```
```suggestion
                            DateTime working_fallback_duration,
```

pbench replied:
Done in 3789955136b424900bec1d4cea2b932e3d6bb971

kinnou02 commented on source/routing/raptor_utils.h:
```diff
@@ -77,6 +77,9 @@ struct Labels {
         dt_pts = clean.dt_pts;
         dt_transfers = clean.dt_transfers;
+
+        walking_duration_pts = clean.walking_duration_pts;
```
same, walking should always be fallback

AurelienLP commented:
Jenkins, test this please

kinnou02 commented on source/routing/raptor_solution_reader.h:
```diff
@@ -53,7 +53,24 @@ struct Dominates {
-typedef ParetoFront<Journey, Dominates> Solutions;
+struct ParetoFrontVisitor {
+    void at_dominated(const Journey& to_insert, const Journey& in_pool) {
+        log4cplus::Logger logger = log4cplus::Logger::getInstance(LOG4CPLUS_TEXT("logger"));
```
You might want to instantiate you logger in the instance, also please name them to help the reading later on

pbench replied:
Ok, the logger is now a member of the struct. https://github.com/CanalTP/navitia/pull/2937/commits/53cd072b2a5f6596daa177532ad871f67da15f12
What do you mean by "name them" ?

nonifier commented on source/routing/benchmark_full.cpp:
```diff
@@ -241,6 +242,22 @@ int main(int argc, char** argv) {
+        std::fstream out_file(demands_output_file, std::ios::out);
+        out_file << "Start id, Target id, , , Day, Hour" << std::endl;
+
+        for (size_t i = 0; i < demands.size(); ++i) {
```
If you want something a little more modern, you can

- Define an output method for `PathDemand`: 

```cpp
std::ostream& operator<<(std::ostream& o, const PathDemand& d) {
    return o << d.start << ", " << d.target << ", , ," << d.date << ", " << d.hour;
}
```
- then use the infamous ostream_iterator :D  
```cpp
boost::copy(demands, std::ostream_iterator<PathDemand>(out_file, "\n"));
```

nonifier commented on source/routing/raptor.cpp:
```diff
@@ -263,11 +285,22 @@ std::vector<StartingPointSndPhase> make_starting_points_snd_phase(const RAPTOR&
+            const unsigned arrival_walking_t = a.second.total_seconds();
+            const DateTime walking_duration_before_arrival_stop_point = working_labels.walking_duration_pt(a.first);
+            const DateTime total_walking_duration = arrival_walking_t + walking_duration_before_arrival_stop_point;
+            StartingPointSndPhase starting_point = {a.first, count,
```
For readability, it would make sense to split the params like:
```cpp
DateTime end_df = working_labels.dt_pt(a.first) + (clockwise ? arrival_walking_t : -arrival_walking_t);
StartingPointSndPhase starting_point = {a.first, count, end_df, total_walking_duration, false};
```

pbench replied:
Yes that's better, done https://github.com/CanalTP/navitia/pull/2937/commits/53cd072b2a5f6596daa177532ad871f67da15f12

nonifier commented on source/routing/raptor.cpp:
```diff
@@ -402,9 +434,16 @@ RAPTOR::Journeys RAPTOR::compute_all_journeys(const map_stop_point_duration& dep
                                               bool clockwise,
                                               const boost::optional<navitia::time_duration>& direct_path_dur,
                                               const size_t max_extra_second_pass) {
+    log4cplus::Logger logger = log4cplus::Logger::getInstance(LOG4CPLUS_TEXT("logger"));
```
Would be interesting to create a new logger (`"raptor"`? maybe) not to pollute the current one and be able to turn it off if it spits out too many things

pbench replied:
Yes, done ! https://github.com/CanalTP/navitia/pull/2937/commits/53cd072b2a5f6596daa177532ad871f67da15f12

nonifier commented on source/routing/raptor.cpp:
```diff
@@ -444,23 +490,44 @@ RAPTOR::Journeys RAPTOR::compute_all_journeys(const map_stop_point_duration& dep
+
         if (solutions.contains_better_than(fake_journey)) {
+            LOG4CPLUS_TRACE(logger, "has better solution than fake journey from " << start_stop_point->uri);
+            for (auto solution : solutions) {
```
```suggestion
            for (const auto& solution : solutions) {
```

pbench replied:
Good catch :) Done https://github.com/CanalTP/navitia/pull/2937/commits/53cd072b2a5f6596daa177532ad871f67da15f12

nonifier commented on source/routing/raptor.cpp:
```diff
@@ -673,16 +749,34 @@ void RAPTOR::raptor_loop(Visitor visitor, const nt::RTLevel rt_level, uint32_t m
                 const auto& jpps_to_explore =
                     visitor.jpps_from_order(data.dataRaptor->jpps_from_jp, jp_idx, q_elt.second);
                 for (const auto& jpp : jpps_to_explore) {
+                    /// jpp : dataRAPTOR::JppsFromJp::Jpp
```
If this comment does help you, you can replace `auto` with the actual type. Comment doesn't usually age well :older_man:  ^^

pbench replied:
Good idea ! Done https://github.com/CanalTP/navitia/pull/2937/commits/53cd072b2a5f6596daa177532ad871f67da15f12

nonifier commented on source/routing/raptor.cpp:
```diff
@@ -794,5 +951,108 @@ int RAPTOR::best_round(SpIdx sp_idx) {
     return -1;
 }
 
+std::string RAPTOR::print_jpps_from_sp() {
```
Are those the following `print_*` functions only for debugging purposes ? 
They don't seem to be used anywhere.

nonifier commented on source/routing/raptor_api.cpp:
```diff
@@ -1363,6 +1370,18 @@ void make_response(navitia::PbCreator& pb_creator,
         return;
     }
 
+    for (const auto stop_point_iter : *departures) {
```
You might not want to iterate through all departures and destinations if you log information above `_TRACE` level.

It would be more interesting to log the whole collections of objects like :
```
  LOG4CPLUS_TRACE(logger, "# departure : " << *departures );
  LOG4CPLUS_TRACE(logger, "# destination : " << *destinations );
```
And you can create the appropriate `ostream& operator<<(...)` function that goes with departures and destinations.

nonifier commented on source/routing/raptor_solution_reader.h:
```diff
@@ -53,7 +53,24 @@ struct Dominates {
-typedef ParetoFront<Journey, Dominates> Solutions;
+struct ParetoFrontVisitor {
+    void at_dominated(const Journey& to_insert, const Journey& in_pool) {
+        log4cplus::Logger logger = log4cplus::Logger::getInstance(LOG4CPLUS_TEXT("logger"));
```
the logger variable could be static to the class and used across the following 3 functions.

pbench replied:
Yep, done https://github.com/CanalTP/navitia/pull/2937/commits/53cd072b2a5f6596daa177532ad871f67da15f12

nonifier commented on source/routing/raptor_solution_reader.h:
```diff
@@ -53,7 +53,24 @@ struct Dominates {
 };
 
-typedef ParetoFront<Journey, Dominates> Solutions;
+struct ParetoFrontVisitor {
```
`ParetoFrontLogger` might be a bit more precise ?

pbench replied:
Done https://github.com/CanalTP/navitia/pull/2937/commits/53cd072b2a5f6596daa177532ad871f67da15f12

nonifier reviewed (approved):
I approve but I don't understand the rational behind the changes ^^

Commit:
taking into account review comments

- used a logger named "raptor" for tracing raptor behavior
instead of the generic one named "logger"
- ParetoFrontVisitor -> ParetoFrontLogger,
- ParetoFrontLogger now has the logger as member of the struct

xlqian commented on source/routing/raptor.cpp:
```diff
@@ -444,23 +490,44 @@ RAPTOR::Journeys RAPTOR::compute_all_journeys(const map_stop_point_duration& dep
+    LOG4CPLUS_TRACE(logger, "starting points 2nd phase " << std::endl
+                                                         << print_starting_points_snd_phase(starting_points));
+
     unsigned lower_bound_fb = std::numeric_limits<unsigned>::max();
```
since it's no longer used, can you remove `lower_bound_fb` ?

pbench replied:
Done 051ab1a4d1bf54e95d2a5d195910e797830ad317

xlqian commented on source/routing/raptor_utils.h:
```diff
@@ -98,6 +112,13 @@ struct Labels {
     IdxMap<type::StopPoint, DateTime> dt_transfers;
+
+    // walking_duration_pts[stop_point] stores the walking duration elapsed during the whole
+    // journey from the starting point which reaches stop point at DateTime dt_pts[stop_point] with public transport
```
does that contain also the fallback part?

xlqian commented on source/routing/raptor_utils.h:
```diff
@@ -98,6 +112,13 @@ struct Labels {
+    // walking_duration_pts[stop_point] stores the walking duration elapsed during the whole
+    // journey from the starting point which reaches stop point at DateTime dt_pts[stop_point] with public transport
+    IdxMap<type::StopPoint, DateTime> walking_duration_pts;
+    // walking_duration_pts[stop_point] stores the walking duration elapsed during the whole
```
```suggestion
    // walking_duration_pts[stop_point] stores the walking transfer duration elapsed during the whole
```
no?

pbougue replied:
No, this is the same as what's done on dt_pts and dt_transfers, but for "fallback" duration (fallback + transfers if I'm correct).

See `PT1` and `TR1` in doc: https://github.com/CanalTP/navitia/blob/dev/documentation/rfc/doc-routing-algorithm.md


```suggestion
    // walking_duration_pts[stop_point] stores the fallback+transfer duration elapsed during the whole
```
(same above)

pbench replied:
For now it only stores the fallback part.
I am testing adding the transfer part, but it has effects on the resulting journeys. 
Not sure yet if it should be just fallback or fallback + transfer.
I added a more precise description comments in 051ab1a4d1bf54e95d2a5d195910e797830ad317

xlqian commented:
there are a lot of commented logs, can you remove or uncomment them?

xlqian commented on source/routing/raptor.cpp:
```diff
@@ -705,11 +808,44 @@ void RAPTOR::raptor_loop(Visitor visitor, const nt::RTLevel rt_level, uint32_t m
-                            next_st->next_stop_time(visitor.stop_event(), jpp.idx, previous_dt, visitor.clockwise());
+                    const DateTime previous_walking_duration = prec_labels.walking_duration_transfer(jpp.sp_idx);
+
+                    /// we are at stop point jpp.idx at time previous_dt
```
can you test `previous_walking_duration <= working_walking_duration` here and `continue` if false? 

it may probably avoid the binary search in `previous_walking_duration <= working_walking_duration`?

benoit-bst commented on source/routing/raptor_api.cpp:
```diff
@@ -1363,6 +1370,18 @@ void make_response(navitia::PbCreator& pb_creator,
         return;
     }
 
+    for (const auto stop_point_iter : *departures) {
```
```suggestion
    for (const auto& stop_point_iter : *departures) {
```

benoit-bst replied:
FYI : Always try to use `const auto&` :cat: to protect fastness and security

pbench replied:
051ab1a4d1bf54e95d2a5d195910e797830ad317

benoit-bst commented on source/routing/raptor_api.cpp:
```diff
@@ -1363,6 +1370,18 @@ void make_response(navitia::PbCreator& pb_creator,
+        LOG4CPLUS_TRACE(logger, "departure : " << stop_point->uri << " distance : " << stop_point_iter.second);
+    }
+
+    for (const auto stop_point_iter : *destinations) {
```
```suggestion
    for (const auto& stop_point_iter : *destinations) {
```

pbench replied:
051ab1a4d1bf54e95d2a5d195910e797830ad317

benoit-bst commented on source/routing/raptor_api.cpp:
```diff
@@ -1363,6 +1370,18 @@ void make_response(navitia::PbCreator& pb_creator,
+
+    for (const auto stop_point_iter : *destinations) {
+        SpIdx sp_idx = stop_point_iter.first;
+        navitia::type::StopPoint* stop_point = raptor.data.pt_data->stop_points[sp_idx.val];
```
```suggestion
        auto stop_point = raptor.data.pt_data->stop_points[sp_idx.val];
```

benoit-bst replied:
Just for good practice :call_me_hand:

pbench replied:
Done 051ab1a4d1bf54e95d2a5d195910e797830ad317

benoit-bst commented on source/routing/raptor_solution_reader.h:
```diff
@@ -53,7 +53,23 @@ struct Dominates {
+    log4cplus::Logger logger = log4cplus::Logger::getInstance(LOG4CPLUS_TEXT("raptor"));
+
+    void at_dominated(const Journey& to_insert, const Journey& in_pool) {
+        LOG4CPLUS_TRACE(logger, "Journey dominated ! " << std::endl
```
Minor : Just for information it's better to use `"\n"` instead of `std::endl` that is more slowly (_endl_ flushes the output buffer)

pbench replied:
Thanks ! I didn't know that

benoit-bst commented on source/routing/raptor.cpp:
```diff
@@ -673,16 +748,33 @@ void RAPTOR::raptor_loop(Visitor visitor, const nt::RTLevel rt_level, uint32_t m
+
                 const auto& jpps_to_explore =
                     visitor.jpps_from_order(data.dataRaptor->jpps_from_jp, jp_idx, q_elt.second);
-                for (const auto& jpp : jpps_to_explore) {
```
why ? `auto` works

pbench replied:
Mostly to help reading. There is so many different objects used that some explicit typing helps (and also help IDE tooltips).

pbougue commented on source/routing/tests/raptor_test.cpp:
```diff
@@ -38,7 +38,13 @@ www.navitia.io
+    logger_initialized() {
+        navitia::init_logger();
+        auto raptor_logger = log4cplus::Logger::getInstance(LOG4CPLUS_TEXT("raptor"));
+        raptor_logger.setLogLevel(log4cplus::FATAL_LOG_LEVEL);
```
Why fatal?

pbench replied:
Just to avoid too much logs on cout during tests

pbougue commented on source/routing/benchmark_full.cpp:
```diff
@@ -241,6 +242,22 @@ int main(int argc, char** argv) {
         }
     }
 
+    // ecriture des requetes
```
:fr: spotted

pbench replied:
Corrected in 78fafeb2b78701bf233a101456b4828701b9e8e4

pbougue commented on source/routing/dataraptor.h:
```diff
@@ -80,6 +80,8 @@ struct dataRAPTOR {
     };
     JppsFromSp jpps_from_sp;
 
+    std::string print_jpps_from_sp();
```
To be removed ?

pbougue commented on source/routing/benchmark_full.cpp:
```diff
@@ -162,6 +162,7 @@ int main(int argc, char** argv) {
             ("verbose,v", "Verbose debugging output")
             ("nb_second_pass", po::value<int>(&nb_second_pass)->default_value(0), "nb second pass")
             ("stop_files", po::value<std::string>(&stop_input_file), "File with list of start and target")
+            ("dump_demands", po::value<std::string>(&demands_output_file), "Write a csv file with the list of demands (start, target, time) used")
```
minor on vocable used in navitia for journeys:
```suggestion
            ("dump_requests", po::value<std::string>(&demands_output_file), "Write a csv file with the list of demands (from, to, datetime) used")
```

pbench replied:
Corrected in 78fafeb2b78701bf233a101456b4828701b9e8e4

pbougue commented on source/routing/benchmark_full.cpp:
```diff
@@ -241,6 +242,22 @@ int main(int argc, char** argv) {
+    // ecriture des requetes
+    if (vm.count("dump_demands")) {
+        std::fstream out_file(demands_output_file, std::ios::out);
+        out_file << "Start id, Target id, , , Day, Hour" << std::endl;
```
Please never split date and time, nowhere!
Well that is unless memory or a data format holds a hostage :smiling_imp: 

(also same remark on vocabulary here)

pbougue commented on source/routing/raptor_solution_reader.cpp:
```diff
@@ -613,9 +621,14 @@ void read_solutions(const RAPTOR& raptor,
+    os << "(["
+       << "departure_dt : " << navitia::str(j.departure_dt) << ", "
+       << "arrival_dt : " << navitia::str(j.arrival_dt) << ", "
+       << "min_wainting_dur : " << j.min_waiting_dur << ", "
```
```suggestion
       << "min_waiting_dur : " << j.min_waiting_dur << ", "
```

pbougue commented on source/routing/raptor_solution_reader.cpp:
```diff
@@ -613,9 +621,14 @@ void read_solutions(const RAPTOR& raptor,
+       << "transfer_dur : " << j.transfer_dur << "], ["
+       << "nb sections : " << j.sections.size() << ", "
+       << "nb extensions : " << unsigned(j.nb_vj_extentions) << "], "
+       << "foot duration : " << j.sn_dur << ")" << std::endl;
```
```suggestion
       << "fallback duration : " << j.sn_dur << ")" << std::endl;
```

pbench replied:
051ab1a4d1bf54e95d2a5d195910e797830ad317

pbougue commented on source/routing/raptor_utils.h:
```diff
@@ -98,6 +112,13 @@ struct Labels {
     IdxMap<type::StopPoint, DateTime> dt_pts;
     // At what time wan we reach this label with a transfer
```
minor
```suggestion
    // At what time can we reach this label with a transfer
```

pbougue commented on source/routing/raptor_utils.h:
```diff
@@ -98,6 +112,13 @@ struct Labels {
+    // journey from the starting point which reaches stop point at DateTime dt_pts[stop_point] with public transport
+    IdxMap<type::StopPoint, DateTime> walking_duration_pts;
+    // walking_duration_pts[stop_point] stores the walking duration elapsed during the whole
+    // journey from the starting point which reaches stop point at DateTime transfers_pts[stop_point] with transfers
```
```suggestion
    // journey from the starting point to 'stop_point' at DateTime transfers_pts[stop_point] with a transfer
```
(same above)

pbench replied:
051ab1a4d1bf54e95d2a5d195910e797830ad317

pbougue commented on source/routing/raptor_solution_reader.h:
```diff
@@ -53,7 +53,23 @@ struct Dominates {
+    }
+};
+
+typedef ParetoFront<Journey, Dominates, ParetoFrontLogger> Solutions;
```
Good for debug, but not sure about perf impact to add function calls in that.
(Same for all logs on highly-used code : maybe better to have a compile-flag?)

pbougue commented on source/routing/raptor.cpp:
```diff
@@ -92,8 +95,15 @@ bool RAPTOR::apply_vj_extension(const Visitor& v,
+                                        << " walk : " << working_walking_duration);
             working_labels.mut_dt_pt(sp_idx) = workingDt;
+            working_labels.mut_walking_duration_pt(sp_idx) = working_walking_duration;
+            BOOST_ASSERT(working_walking_duration != DateTimeUtils::not_valid);
```
minor: Why not a "simple" `assert` ?
You might want to properly raise a recoverable exception if you keep it, too.

> :heavy_check_mark: OK, we already have some others, that's good like that!

pbougue commented on source/routing/raptor.cpp:
```diff
@@ -62,7 +62,10 @@ bool RAPTOR::apply_vj_extension(const Visitor& v,
                                 const uint16_t l_zone,
-                                DateTime base_dt) {
+                                DateTime base_dt,
+                                DateTime working_walking_duration,
```
Changing signature just for logs is weird.
Sorry, wrong line, it's :point_down:

pbougue commented on source/routing/raptor.cpp:
```diff
@@ -285,19 +317,18 @@ std::vector<StartingPointSndPhase> make_starting_points_snd_phase(const RAPTOR&
     Journey journey;
     journey.sections.resize(sp.count);  // only the number of sections is part of the dominance function
-    journey.sn_dur = navitia::time_duration(0, 0, sp.fallback_dur + lower_bound_fb, 0);
+    journey.sn_dur = navitia::time_duration(0, 0, sp.fallback_dur, 0);
```
Are you sure it's the minimal value ?
With the possible "wrong update" example we discussed, can't we end up with a too high lower bound (stopping the search too early)?

pbougue commented on source/routing/raptor.cpp:
```diff
@@ -285,19 +317,18 @@ std::vector<StartingPointSndPhase> make_starting_points_snd_phase(const RAPTOR&
     if (clockwise) {
         journey.arrival_dt = sp.end_dt;
-        journey.departure_dt = sp.end_dt - journey.sn_dur.seconds() - nb_conn * lower_bound_conn;
+        journey.departure_dt = sp.end_dt - journey.sn_dur.seconds();
```
:+1: OK, already managed in "new" sn_dur.

pbougue commented on source/routing/raptor.cpp:
```diff
@@ -673,16 +748,33 @@ void RAPTOR::raptor_loop(Visitor visitor, const nt::RTLevel rt_level, uint32_t m
+
             const JpIdx jp_idx = q_elt.first;
+
+            const RouteIdx route_idx = data.dataRaptor->jp_container.get(jp_idx).route_idx;
```
Please create vars as close as possible from their use :)

pbougue reviewed (commented):
Can you update https://github.com/CanalTP/navitia/blob/dev/documentation/rfc/doc-routing-algorithm.md eventually?

I'd love to have some insight of what you changed in that loop tomorrow, coz it's hard to follow : https://github.com/CanalTP/navitia/pull/2937/files?file-filters%5B%5D=.cpp#diff-30631442826dc5f74f14e880e96c74bcR777 :wink:

pbougue commented on source/routing/raptor.cpp:
```diff
@@ -705,11 +808,44 @@ void RAPTOR::raptor_loop(Visitor visitor, const nt::RTLevel rt_level, uint32_t m
+                    //  hence we can board any vehicle arriving after previous_dt
                     const DateTime previous_dt = prec_labels.dt_transfer(jpp.sp_idx);
-                    if (prec_labels.transfer_is_initialized(jpp.sp_idx) && valid_stop_points[jpp.sp_idx.val]
-                        && (!is_onboard || visitor.better_or_equal(previous_dt, base_dt, *it_st))) {
```
So `better_or_equal` is not used anymore... We should probably remove it if it stays that way.
I don't understand what this tested?

pbougue commented on source/routing/raptor.cpp:
```diff
@@ -705,11 +808,44 @@ void RAPTOR::raptor_loop(Visitor visitor, const nt::RTLevel rt_level, uint32_t m
+
+                    // if we cannot board at this stop point, nothing to do
+                    if (!prec_labels.transfer_is_initialized(jpp.sp_idx) || !valid_stop_points[jpp.sp_idx.val]) {
+                        continue;
```
If `is_onboard==true`, shouldn't we do the stay_in check https://github.com/CanalTP/navitia/pull/2937/files#diff-30631442826dc5f74f14e880e96c74bcR892 like before or is it pointless ?

pbougue commented on source/routing/raptor.cpp:
```diff
@@ -705,11 +808,44 @@ void RAPTOR::raptor_loop(Visitor visitor, const nt::RTLevel rt_level, uint32_t m
+
+                    /// if there is no vehicle arriving after previous_dt, nothing to do
+                    if (tmp_st_dt.first == nullptr) {
+                        continue;
```
Same, what about the stay_in check?

Commit:
cleaning benchmark_full.cpp

- Input and output date_times are now in posix_time.
- French -> English
- more precise --help message

Commit:
renamed walking_duration to fallback_duration

Commit:
comments and cleaning

Commit:
enhance logging output

Commit:
store best fallback duration along with best arrival time

Commit:
fixed initialization of best walking duration

Commit:
always launch a second pass from arrival stop point which have priority

Commit:
test total waiting dur

Commit:
initialize best labels with foot path duration

Commit:
removed transfer penalty from bound journey  in solution reader

Commit:
updated journey comparator for pareto front

Commit:
Fixing raptor_tests

The _bound_ in raptor.compute() was interpreted before as "a journey must arrive strictly before  _bound_". The bound now means "a journey must arrive before or at _bound_".
The tests are updated accordingly

Commit:
total computing time printed by benchmark_full

Commit:
use transfer_penalty in journey comparison

instead of comparing just the walking time, we compare walking time + transfer_penalty * nb_of_sections
so that a journey with more sections but less walking is discarded if the additional section does not allows to reduce the walking duration by more than transfer_penalty

Commit:
update tests

The _bound_ in raptor.compute(..., clockwise = false) was interpreted before as
"a journey must depart strictly after _bound_".
It now means "a journey must depart after or at _bound_".

Commit:
fix compil after rebase

Commit:
cleaning not so usefull logs and comments

Commit:
use the total walking duration in first phase instead of only fallback duration

Commit:
transfer penalty correctly used

Commit:
when trying to change board point, use arrival time instead of departure time for candidate

Commit:
adding a walking transfer penalty of 30s

Commit:
update computation of the next kraken call datetime

Commit:
fix next kraken call datetime computation

Commit:
if arrival times are the same, compare penalized departures

Commit:
jormun filter : compare departure times when arrival times are identical

Commit:
remove total waiting duration from journey comparison criteria

Commit:
1m30 for walking transfer penalty

Commit:
more precise walking duration in fake journeys

Commit:
deactyivate logs

Commit:
fix raptor test

Commit:
2mn walking penalty

Commit:
fixing tests

Commit:
fixing another test

Commit:
another test fixed

Commit:
more tests fixed

Commit:
fixing remaining failing tests

Commit:
remove logs in pareto front updates

Commit:
cleaning comments and logs

Commit:
fixing yet another test
